### PR TITLE
[move core modules] events for mint, burn, preburn, cancel_burn

### PR DESCRIPTION
**Pull request by sblackshear**
Getting closer to production-like mint/burn flows!
    
- Add events for each of `mint`, `burn`, `preburn`, and `cancel_burn`. These events are only emitted for reserve-backed stablecoins (i.e., `Coin1`, `Coin2` for now), not for synthetic coins like `LBR`. The reason is that creating/destroying reserve-backed coins is a privileged operation, but anyone can create/destroy `LBR`.
- Add a `BurnCapability` (allocated to the `0xA550C18` account, just like the `MintCapability`) and require it to burn. Previously, the holder of the `MintCapability` also had permission to burn.
    
To make this work, I had to make some unrelated tweaks:
- The genesis setup initializes the `CurrencyInfo` for `Coin1` and `Coin2` (which now require event streams created by `0xA550C18`) before it creates the `0xA550C18` account. It needs to be this way because account creation depends on `Coin1` and `Coin2` existing... To fix this, I added somewhat of a hack to give `0xA550C18` the ability to create event streams before its account is created. This hack is limited to two local changes in the Event module that can only be used in genesis. Tim and I brainstormed several ways to mitigate this and concluded that this one is ugly, but not fragile (unlike the other solutions we considered). Added a TODO for thinking of a better approach sometime in the future...
- All of the existing mint/burn/preburn/cancel_burn tests used `LBR`. I switched them to use `Coin1` because minting/burning of `LBR` isn't privileged and doesn't emit events.
- I had to add max-gas and gas-price directives to many of the functional tests I touched because gas costs are now much higher and we don't yet support collecting tx fees in non-`LBR` currencies.

## Motivation

See above.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

New tests to check that events are emitted at the appropriate times.

### CODE REVIEW

**Comment by dmitri-perelman** [comment]
Awesome! (side note: it's safe to assume that LBR cannot be minted or burned, so it's absolutely ok to build these for reserve-backed stablecoins only)

**Comment by andll** [comment]
`0xA550C18` is a generic 'association account', right?
Can we create separate account in genesis for burning and give association account capability to rotate its key?
Separate account is needed because we will rotate key of this account to the key of HSM (and probably don't want to rotate master association key, as it is used for minting and reconfiguration which is not yet supported in HSM)

File: language/tools/vm-genesis/src/lib.rs [hunk 1]
```diff
@@ -291,6 +291,17 @@ fn create_and_initialize_main_accounts(
...
+            .execute_function(
+                &module("Event"),
+                &name("initialize"),
+                &gas_schedule,
+                interpreter_context,
+                &txn_data,
+                vec![],
+                vec![],
+            )
+            .expect("Failure initializing currency module");
```
Comment (tzakian) [inline language/tools/vm-genesis/src/lib.rs:304]: Nit: update the string in this expect

File: types/src/on_chain_config/mod.rs [hunk 2]
```diff
@@ -122,7 +122,7 @@ pub trait OnChainConfig: Send + Sync + DeserializeOwned {
 }
 
 pub fn new_epoch_event_key() -> EventKey {
-    EventKey::new_from_address(&association_address(), 4)
+    EventKey::new_from_address(&association_address(), 16)
```
Comment (tzakian) [inline types/src/on_chain_config/mod.rs:125]: Is there any particular reason for this change? I'm more curious than anything else :)

File: language/stdlib/modules/lbr.move [hunk 3]
```diff
@@ -94,6 +97,7 @@ module LBR {
         let coin2_exact = Libra::withdraw(&mut coin2, num_coin2);
         Libra::deposit(&mut reserve.coin1.backing, coin1_exact);
         Libra::deposit(&mut reserve.coin2.backing, coin2_exact);
+        // TODO: This will emit an event. Do we want that for LBR?
```
Comment (tzakian) [inline language/stdlib/modules/lbr.move:100]: I think this TODO was resolved in this PR, so can probably remove?

**Comment by sblackshear** [comment]
> Can we create separate account in genesis for burning and give association account capability to rotate its key?

Yep, that makes a ton of sense. Now that we have a `BurnCapability`, we can change the genesis setup to
- create a burn account (eager to hear creative hex suggestions for address. 0xD1E is the best I can come up with)
- give the `BurnCapability<CoinType>` for each `CoinType` to `0xD1E`
- delegate `0xD1E`'s `KeyRotationCapability` to `A550C18`

once this PR lands.

File: types/src/on_chain_config/mod.rs (hunk 2 above)
Comment (sblackshear) [reply types/src/on_chain_config/mod.rs:125]: Ahh great question. The association now creates 12 new event handles before creating the `new_epoch` handle: 4 each (burn event, preburn event, cancel burn event, mint event) for the three currencies.

File: language/stdlib/modules/lbr.move (hunk 3 above)
Comment (sblackshear) [reply language/stdlib/modules/lbr.move:100]: Good catch. I wrote the TODO and then discovered the magic of `is_synthetic`. Whoever introduced that is a genius.

**Comment by andll** [comment]
I can take a stub at creating separate account with your help :)

**Comment by dmitri-perelman** [comment]
Totally unrelated question: is there some hidden rule behind the special addresses like `0xD1E` or `0xA550C18`? (I was playing with kabbalah numerology for these but didn't find anything)

**Comment by sblackshear** [comment]
> Totally unrelated question: is there some hidden rule behind the special addresses like `0xD1E` or `0xA550C18`? (I was playing with kabbalah numerology for these but didn't find anything)

Haha no hidden rule other than trying to write descriptive human-readable addresses in hex... @phlip9's `0xD15C0` for the discovery set is my favorite so far

**Pull request closed without merging**